File: solutions/738_monotone_increasing_digits.rb
#
# Given a non-negative integer N, find the largest number that is less than or equal to N with monotone increasing digits.
# (Recall that an integer has monotone increasing digits if and only if each pair of adjacent digits x and y satisfy x <= y.)
# Example 1:
# Input: N = 10
# Output: 9
# Example 2:
# Input: N = 1234
# Output: 1234
# Example 3:
# Input: N = 332
# Output: 299
# Note:
# N is an integer in the range [0, 10^9].
#

# @param {Integer} n
# @return {Integer}
def monotone_increasing_digits(n)
  nstr = n.to_s.chars.map(&:to_i)
  loc = nstr.size + 1
  (nstr.size - 1).downto(1).each do |i|
    next if nstr[i] >= nstr[i - 1]
    nstr[i - 1] -= 1
    loc = i
  end
  loc.upto(nstr.size - 1).each { |i| nstr[i] = '9' }
  nstr.join.to_i
end

n = 5_666_661
p monotone_increasing_digits(n)

File: _site/solutions/738_monotone_increasing_digits.cpp
#include "aux.cpp"
#include <algorithm>
#include <climits>
#include <iostream>
#include <map>
#include <set>
#include <stdio.h>
#include <unordered_map>
#include <vector>
/**

Given a non-negative integer N, find the largest number that is less than or
equal to N with monotone increasing digits. (Recall that an integer has monotone
increasing digits if and only if each pair of adjacent digits x and y satisfy x
<= y.) Example 1: Input: N = 10 Output: 9 Example 2: Input: N = 1234 Output:
1234 Example 3: Input: N = 332 Output: 299 Note: N is an integer in the range
[0, 10^9].

 **/
using namespace std;

static const int _ = []() {
  ios::sync_with_stdio(false);
  cin.tie(NULL);
  return 0;
}();
class Solution {
public:
  int monotoneIncreasingDigits(int N) {
    string n_str = to_string(N);
    int loc = n_str.size(), res = 0;
    for (int i = n_str.size() - 1; i > 0; i--)
      if (n_str[i] < n_str[i - 1]) {
        loc = i;
        n_str[i - 1] -= 1;
      }
    for (int i = loc; i < n_str.size(); i++)
      n_str[i] = '9';

    return stoi(n_str);
  }
};

int main() {
  Solution s;
  int n = 1234;
  say(s.monotoneIncreasingDigits(n));

  srand(time(NULL));
  n = rand() % 100000;
  say(n);
}
